components/support/nimbus-fml/src/structure/validator.rs

```diff
@@ -0,0 +1,485 @@
+                    )?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn validate_type_ref(&self, path: &str, type_ref: &TypeRef) -> Result<()> {
+        match type_ref {
+            TypeRef::Enum(name) => {
+                if !self.enum_defs.contains_key(name) {
+                    return Err(FMLError::ValidationError(
+                        path.to_string(),
+                        format!("Found enum reference with name: {name}, but no definition"),
+                    ));
+                }
+            }
+            TypeRef::Object(name) => {
+                if !self.object_defs.contains_key(name) {
+                    return Err(FMLError::ValidationError(
+                        path.to_string(),
+                        format!("Found object reference with name: {name}, but no definition"),
+                    ));
+                }
+            }
+            TypeRef::EnumMap(key_type, value_type) => match key_type.as_ref() {
+                TypeRef::Enum(_) | TypeRef::String | TypeRef::StringAlias(_) => {
+                    self.validate_type_ref(path, key_type)?;
+                    self.validate_type_ref(path, value_type)?;
+                }
+                _ => {
+                    return Err(FMLError::ValidationError(
+                        path.to_string(),
+                        format!(
+                            "Map key must be a String, string-alias or enum, found: {key_type:?}",
+                        ),
+                    ))
+                }
+            },
+            TypeRef::List(list_type) => self.validate_type_ref(path, list_type)?,
+            TypeRef::StringMap(value_type) => self.validate_type_ref(path, value_type)?,
+            TypeRef::Option(option_type) => {
+                if let TypeRef::Option(_) = option_type.as_ref() {
+                    return Err(FMLError::ValidationError(
+                        path.to_string(),
+                        "Found nested optional types".into(),
+                    ));
+                } else {
+                    self.validate_type_ref(path, option_type)?
+                }
+            }
+            _ => (),
+        };
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod manifest_structure {
+    use serde_json::json;
+
+    use super::*;
+    use crate::error::Result;
+    use crate::intermediate_representation::PropDef;
+
+    #[test]
+    fn validate_enum_type_ref_doesnt_match_def() -> Result<()> {
+        let enums = Default::default();
+        let objs = Default::default();
+        let validator = StructureValidator::new(&enums, &objs);
+        let fm = FeatureDef::new(
+            "some_def",
+            "test doc",
+            vec![PropDef::new(
+                "prop name",
+                TypeRef::Enum("EnumDoesntExist".into()),
+                json!(null),
+            )],
+            false,
+        );
+        validator.validate_feature_def(&fm).expect_err(
```

Comment:
All these tests used to be of the form:

```
let fm = get_feature_manifest(…)

fm.validate_something()
```

I've left most of the variable names in tact.